File: backend/app/security.py
import uuid
from collections import defaultdict, deque
from datetime import datetime, timedelta, timezone

import bcrypt
import jwt

from app.config import settings

APP_PASSWORD_HASH: str | None = None

_login_attempts: dict[str, deque[datetime]] = defaultdict(deque)
LOGIN_WINDOW = timedelta(seconds=60)
LOGIN_MAX_ATTEMPTS = 5
_MAX_IPS = 10_000

_revoked_jtis: dict[str, datetime] = {}


def hash_password(password: str) -> str:
    return bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt()).decode("ascii")


def verify_password(password: str, password_hash: str) -> bool:
    try:
        return bcrypt.checkpw(password.encode("utf-8"), password_hash.encode("ascii"))
    except ValueError:
        return False


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


def check_login_rate(ip: str) -> bool:
    now = utc_now()
    queue = _login_attempts[ip]
    while queue and now - queue[0] > LOGIN_WINDOW:
        queue.popleft()
    if len(queue) >= LOGIN_MAX_ATTEMPTS:
        return False
    if len(_login_attempts) >= _MAX_IPS and ip not in _login_attempts:
        for key in list(_login_attempts.keys()):
            queue_k = _login_attempts[key]
            if not queue_k or (now - queue_k[-1] > LOGIN_WINDOW):
                del _login_attempts[key]
            if len(_login_attempts) < _MAX_IPS:
                break
    queue.append(now)
    return True


def create_access_token(device_id: uuid.UUID) -> str:
    now = utc_now()
    jti = uuid.uuid4()
    return jwt.encode(
        {
            "sub": str(device_id),
            "type": "access",
            "jti": str(jti),
            "iat": now,
            "exp": now + timedelta(minutes=settings.access_token_minutes),
        },
        settings.jwt_secret,
        algorithm="HS256",
    )


def create_refresh_token(device_id: uuid.UUID, jti: uuid.UUID) -> str:
    now = utc_now()
    return jwt.encode(
        {
            "sub": str(device_id),
            "type": "refresh",
            "jti": str(jti),
            "iat": now,
            "exp": now + timedelta(days=settings.refresh_token_days),
        },
        settings.jwt_secret,
        algorithm="HS256",
    )


def _decode_access_payload(token: str) -> dict:
    payload = jwt.decode(token, settings.jwt_secret, algorithms=["HS256"])
    if payload.get("type") != "access":
        raise jwt.InvalidTokenError("not an access token")
    if payload.get("jti") in _revoked_jtis:
        raise jwt.InvalidTokenError("access token revoked")
    return payload


def decode_access_token(token: str) -> uuid.UUID:
    return uuid.UUID(_decode_access_payload(token)["sub"])


def decode_access_token_jti(token: str) -> str:
    return str(_decode_access_payload(token)["jti"])


def decode_refresh_token(token: str) -> tuple[uuid.UUID, uuid.UUID]:
    payload = jwt.decode(token, settings.jwt_secret, algorithms=["HS256"])
    if payload.get("type") != "refresh":
        raise jwt.InvalidTokenError("not a refresh token")
    return uuid.UUID(payload["sub"]), uuid.UUID(payload["jti"])


def revoke_access_tokens(*jtis: str) -> None:
    now = utc_now()
    for jti in jtis:
        _revoked_jtis[jti] = now


def cleanup_revoked_jtis(max_age_seconds: int = 900) -> int:
    now = utc_now()
    cutoff = timedelta(seconds=max_age_seconds)
    expired = [jti for jti, revoked_at in _revoked_jtis.items() if now - revoked_at > cutoff]
    for jti in expired:
        _revoked_jtis.pop(jti, None)
    return len(expired)

File: backend/internal/security/security.go
package security

import (
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
)

var (
	ErrInvalidToken       = errors.New("invalid token")
	ErrTokenRevoked       = errors.New("token revoked")
	ErrNotAccessToken     = errors.New("not an access token")
	ErrNotRefreshToken    = errors.New("not a refresh token")
	ErrInvalidCredentials = errors.New("invalid credentials")
)

type SecurityManager struct {
	jwtSecret           []byte
	accessTokenDuration time.Duration
	refreshTokenDays    int
	passwordHash        string

	// Login Rate Limiter
	rateMu        sync.Mutex
	loginAttempts map[string][]time.Time
	maxIPs        int

	// In-memory revoked JTIs for access tokens
	jtiMu       sync.RWMutex
	revokedJTIs map[string]time.Time
}

func NewSecurityManager(jwtSecret string, accessMinutes, refreshDays int, passwordHash string) *SecurityManager {
	return &SecurityManager{
		jwtSecret:           []byte(jwtSecret),
		accessTokenDuration: time.Duration(accessMinutes) * time.Minute,
		refreshTokenDays:    refreshDays,
		passwordHash:        passwordHash,
		loginAttempts:       make(map[string][]time.Time),
		maxIPs:              10000,
		revokedJTIs:         make(map[string]time.Time),
	}
}

func (s *SecurityManager) SetPasswordHash(hash string) {
	s.passwordHash = hash
}

func (s *SecurityManager) GetPasswordHash() string {
	return s.passwordHash
}

func HashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}

func (s *SecurityManager) VerifyPassword(password string) bool {
	if s.passwordHash == "" {
		return false
	}
	err := bcrypt.CompareHashAndPassword([]byte(s.passwordHash), []byte(password))
	return err == nil
}

func (s *SecurityManager) CheckLoginRate(ip string) bool {
	s.rateMu.Lock()
	defer s.rateMu.Unlock()

	now := time.Now().UTC()
	window := 60 * time.Second
	maxAttempts := 5

	// Clean expired for this IP
	attempts := s.loginAttempts[ip]
	validAttempts := attempts[:0]
	for _, t := range attempts {
		if now.Sub(t) <= window {
			validAttempts = append(validAttempts, t)
		}
	}

	if len(validAttempts) >= maxAttempts {
		s.loginAttempts[ip] = validAttempts
		return false
	}

	// Clean map if exceeding maxIPs
	if len(s.loginAttempts) >= s.maxIPs && len(validAttempts) == 0 {
		for k, v := range s.loginAttempts {
			if len(v) == 0 || now.Sub(v[len(v)-1]) > window {
				delete(s.loginAttempts, k)
			}
			if len(s.loginAttempts) < s.maxIPs {
				break
			}
		}
	}

	validAttempts = append(validAttempts, now)
	s.loginAttempts[ip] = validAttempts
	return true
}

func (s *SecurityManager) CreateAccessToken(deviceID uuid.UUID) (string, error) {
	now := time.Now().UTC()
	jti := uuid.New()
	claims := jwt.MapClaims{
		"sub":  deviceID.String(),
		"type": "access",
		"jti":  jti.String(),
		"iat":  now.Unix(),
		"exp":  now.Add(s.accessTokenDuration).Unix(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(s.jwtSecret)
}

func (s *SecurityManager) CreateRefreshToken(deviceID, jti uuid.UUID) (string, error) {
	now := time.Now().UTC()
	claims := jwt.MapClaims{
		"sub":  deviceID.String(),
		"type": "refresh",
		"jti":  jti.String(),
		"iat":  now.Unix(),
		"exp":  now.Add(time.Duration(s.refreshTokenDays) * 24 * time.Hour).Unix(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(s.jwtSecret)
}

func (s *SecurityManager) DecodeAccessToken(tokenStr string) (uuid.UUID, string, error) {
	token, err := jwt.Parse(tokenStr, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return s.jwtSecret, nil
	})
	if err != nil || !token.Valid {
		return uuid.Nil, "", ErrInvalidToken
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return uuid.Nil, "", ErrInvalidToken
	}

	tokenType, _ := claims["type"].(string)
	if tokenType != "access" {
		return uuid.Nil, "", ErrNotAccessToken
	}

	jti, _ := claims["jti"].(string)
	if jti == "" {
		return uuid.Nil, "", ErrInvalidToken
	}

	s.jtiMu.RLock()
	_, revoked := s.revokedJTIs[jti]
	s.jtiMu.RUnlock()
	if revoked {
		return uuid.Nil, "", ErrTokenRevoked
	}

	subStr, _ := claims["sub"].(string)
	deviceID, err := uuid.Parse(subStr)
	if err != nil {
		return uuid.Nil, "", ErrInvalidToken
	}

	return deviceID, jti, nil
}

func (s *SecurityManager) DecodeRefreshToken(tokenStr string) (uuid.UUID, uuid.UUID, error) {
	token, err := jwt.Parse(tokenStr, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return s.jwtSecret, nil
	})
	if err != nil || !token.Valid {
		return uuid.Nil, uuid.Nil, ErrInvalidToken
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return uuid.Nil, uuid.Nil, ErrInvalidToken
	}

	tokenType, _ := claims["type"].(string)
	if tokenType != "refresh" {
		return uuid.Nil, uuid.Nil, ErrNotRefreshToken
	}

	subStr, _ := claims["sub"].(string)
	deviceID, err := uuid.Parse(subStr)
	if err != nil {
		return uuid.Nil, uuid.Nil, ErrInvalidToken
	}

	jtiStr, _ := claims["jti"].(string)
	jti, err := uuid.Parse(jtiStr)
	if err != nil {
		return uuid.Nil, uuid.Nil, ErrInvalidToken
	}

	return deviceID, jti, nil
}

func (s *SecurityManager) RevokeAccessTokens(jtis ...string) {
	s.jtiMu.Lock()
	defer s.jtiMu.Unlock()
	now := time.Now().UTC()
	for _, jti := range jtis {
		if jti != "" {
			s.revokedJTIs[jti] = now
		}
	}
}

func (s *SecurityManager) CleanupRevokedJTIs(maxAge time.Duration) int {
	s.jtiMu.Lock()
	defer s.jtiMu.Unlock()

	now := time.Now().UTC()
	removed := 0
	for jti, revokedAt := range s.revokedJTIs {
		if now.Sub(revokedAt) > maxAge {
			delete(s.revokedJTIs, jti)
			removed++
		}
	}
	return removed
}

func (s *SecurityManager) ResetForTesting() {
	s.rateMu.Lock()
	s.loginAttempts = make(map[string][]time.Time)
	s.rateMu.Unlock()

	s.jtiMu.Lock()
	s.revokedJTIs = make(map[string]time.Time)
	s.jtiMu.Unlock()
}

